add support for html-webpack-plugin

ceonizm opened the pull request:
## Description
This PR adds support for html-webpack-plugin as part of issue #36 
it allows to inject the sprites generated into the html template(s) generated by html-webpack-plugin
I've added the following options: 

- injectSpritesInTemplates : allows to enable the injection (can also be an object that allows to fine tune more the injection (specify the position and entries to operate)

alexander-akait commented on src/index.ts:
```diff
@@ -200,6 +209,66 @@ class SvgChunkWebpackPlugin {
+			const HtmlWebpackPlugin = HtmlWebpackPluginImport.default;
+			const hooks = HtmlWebpackPlugin.getCompilationHooks(compilation);
+
+			if (this.options.injectSpritesInTemplates) {
```
I recommend to move `try/catch` into `if`, because you don't need to make `import`, i.e. `fs`/`network` call if option is disabled

yoriiis commented:
@ceonizm About lint errors, you can use the VSCode workspace which will automatically format the code. Or use the npm script `npm run biome:fix`.

yoriiis commented on README.md:
```diff
@@ -257,6 +257,41 @@ Default: `false`
+
+Default: `false`
+
+Allows the plugin to inject the sprites into the html files generated by [html-webpack-plugin](https://github.com/jantimon/html-webpack-plugin).
```
```suggestion
Allows the plugin to inject the sprites into the HTML files generated by [html-webpack-plugin](https://github.com/jantimon/html-webpack-plugin).
```

yoriiis commented on src/schemas/plugin-options.json:
```diff
@@ -90,6 +90,34 @@
+				{
+					"type": "boolean"
+				},
+				{
+					"type": "object",
+					"properties": {
+						"position": {
+							"link": "https://github.com/yoriiis/svg-chunk-webpack-plugin#injectSpritesInTemplates.position",
+							"description": "the position where to inject the sprites map: either at the beginning of the body or at the end",
+							"type": "string",
+							"enum": ["before", "after"]
+						},
+						"entries": {
+							"link": "https://github.com/yoriiis/svg-chunk-webpack-plugin#injectSpritesInTemplates.entries",
+							"description": "a list of entries we want to include the sprites on (in case of multiples entries generated by htmlWebpackPlugin",
+							"type": "array",
+							"items": {
+								"type": "string"
+							}
+						}
+					}
+				}
```
Is it really necessary to make the injection position customizable? Before closing the body tag seems sufficient by default, doesn't it?

ceonizm replied:
> Is it really necessary to make the injection position customizable? Before closing the body tag seems sufficient by default, doesn't it?

I agree that it should be sufficient for most cases (in fact I don't even see what case may require to get the sprite-map at the bottom of body). For the record I started by using the htmlwebpackplugin api and pushing the tag into  bodytags then noticed it was adding to the end so I did the other part of code :)

yoriiis replied:
On HtmlWebpackPlugin, for styles or scripts, it's interesting to be able to choose the position (and even then, only for scripts). But for our purposes, an injection at the bottom of the DOM before closing the `</body>` tag seems sufficient. It should also simplify your code

ceonizm replied:
In fact sprites seems to work better when inserted at the first (eg: just after the opening `<body>` tag).

yoriiis replied:
Okay, so can we just keep the `before` injection as the default behavior and remove the customization?

yoriiis commented on tests/index.test.js:
We'll have to test the changes added in this PR. Start only when all the feedback is ok to avoid rewriting the tests

ceonizm commented on src/index.ts:
```diff
@@ -200,6 +211,66 @@ class SvgChunkWebpackPlugin {
+			const hooks = HtmlWebpackPlugin.getCompilationHooks(compilation);
+
+			if (this.options.injectSpritesInTemplates) {
+				hooks.afterTemplateExecution.tapAsync({name: 'SvgChunkWebpackPlugin'}, (data, callback) => {
```
ok got it

yoriiis commented on src/index.ts:
```diff
@@ -200,6 +211,66 @@ class SvgChunkWebpackPlugin {
+							if (insertPosition === 'before') {
+								data.html = data.html.replace(/<body([^>]*)>/i, `<body$1>${sprite.sprite}`);
+							} else {
+								const res= sprite.sprite.match(/^<svg([^>]*)?>/i);
+								const attributes:Record<string,string> = {};
+								if( res && res[1]) {
+									const re = /(?<property>[\w|-]+)=["']?(?<value>(?:.(?!["']?\s+(?:\S+)=|\s*\/?[>"']))+.)["']?/gi;
+									let attr
+									while(  attr = re.exec(res[1])) {
+										if (attr && attr.groups) {
+											attributes[attr.groups.property] = attr.groups.value;
+										}
+									}
+								}
+								data.bodyTags.unshift(
+									{
+										tagName: 'svg',
+										voidTag: false,
+										meta: {
+											plugin: 'svg-chunk-webpack-plugin'
+										},
+										innerHTML: sprite.sprite.replace(/<svg[^>]*?>([^$]+?)<\/svg>/mgi, '$1'),
+										attributes: attributes
+									}
+								);
```
This part is complex, so why can't it be as simple as the `before` case?

ceonizm replied:
> This part is complex, so why can't it be as simple as the `before` case?

because I use the HtmlWebpackPlugin API which required me to define the svg tag as one of their HtmlTagObject so it forces me to 

- extract attributes from the already generated markup to insert it the attributes object
- extract the content of the svg markup to add it as innerHtml property

for sure I may have been ugly and just do a string replace it would have been way easier :)

yoriiis replied:
Question, do we really gain by using the HtmlWebpackPlugin API? Natively with the webpack API it wouldn't be feasible/simpler (I'm not familiar with that part).

Okay but why is injection simpler in the `before` case? (one line compare to +20 lines)

ceonizm replied:
> Okay but why is injection simpler in the before case? (one line compare to +20 lines)
Simply because as it is not handled by HtmlWebpackPlugin API I did a simple and "ugly" string replace.

Just made a push that did the same for `after` case.

ceonizm commented:
> @ceonizm About lint errors, you can use the VSCode workspace which will automatically format the code. Or use the npm script `npm run biome:fix`.

not a big fan of biome answers about lint/complexity/useOptionalChain

yoriiis commented:
> Hello @yoriiis Regarding this linter message, https://github.com/yoriiis/svg-chunk-webpack-plugin/actions/runs/14497563068/job/40733325804?pr=37#step:5:16
> 
> Sorry but I don't consider it as an error by myself. Since Biome themselves state their "fix" as "unsafe" I think this rule should be disabled.

Biome continues to recommend this rule. I don't know if it's the automatic fix that's unsafe or the change itself. The plugin requires a recent version of Node.js (+14) and optional chaining is supported

yoriiis commented on src/index.ts:
```diff
@@ -200,6 +208,68 @@ class SvgChunkWebpackPlugin {
+		let HtmlWebpackPluginImport: any;
+		try {
+			// @ts-ignore
+			HtmlWebpackPluginImport = await import('html-webpack-plugin');
```
The import is performed each time the function is called. Could the import be done upstream?

cc/ @alexander-akait

yoriiis replied:
Can we imagine a function that returns the instance and stores it as a class property so that it can be imported just once?

I think @alexander-akait has already given me feedback on the performance issue of storing this type of data as a class property

yoriiis replied:
I get the instance from the webpack compiler options

yoriiis commented:
@ceonizm I've reworked the code and added a test in the demo. Let me know what you think. If it's OK, I'll update the doc and then we'll see about the tests.

ceonizm commented:
> @ceonizm I've reworked the code and added a test in the demo. Let me know what you think. If it's OK, I'll update the doc and then we'll see about the tests.

The idea of gathering the plugin directly from the compilation object instead of doing another import is pretty nice and got the advantage in bonus to avoid the question of using or not that optional chain syntax :)

I'm just not totally sure to understand the reason of this: 
```const spriteByEntries = data.plugin.options.chunks.map((entry) =>
			sprites.find((sprite) => sprite.entryName.includes(entry))
		);
```

Naively I thought the webpack plugin emitted each entry file only once and the data.outputName was enough for that purpose.

Otherwise you're gonna laugh but ... I was thinking about it on my lunch time: since we finally don't care about the html-webpack-plugin api and end with a "savage" string replacement, I think there is finally an easier approach which would consist of  not relying at all on html-webpack-plugin and just tap the standard `assetEmitted` webpack hook. This way it would be even compatible with any plugins that would produce html files.

yoriiis commented:
> Naively I thought the webpack plugin emitted each entry file only once and the data.outputName was enough for that purpose.
> 
> Otherwise you're gonna laugh but ... I was thinking about it on my lunch time: since we finally don't care about the html-webpack-plugin api and end with a "savage" string replacement, I think there is finally an easier approach which would consist of not relying at all on html-webpack-plugin and just tap the standard `assetEmitted` webpack hook. This way it would be even compatible with any plugins that would produce html files.

Do you know of any other plugins that do the same thing? I have the impression that this is the most popular. It's true that ideally I'd prefer not to create a link with my plugin because that would require maintenance/doc/monitoring etc.

If you only use the Webpack API, I don't see how you can detect potential templates associated with entrypoints. We could just replace the hook part extracted from HtmlWebpackPlugin with the native webpack hooks.

yoriiis commented on src/index.ts:
```diff
@@ -200,6 +207,55 @@ class SvgChunkWebpackPlugin {
+	 * @param {Compilation} compilation Webpack compilation
+	 * @returns HtmlWebpackPlugin instance or null
+	 */
+	getHtmlWebpackPluginInstance(compilation: Compilation): any | null {
```
Is it possible to replace the `any` type with a more qualified type?

ceonizm commented:
> > Naively I thought the webpack plugin emitted each entry file only once and the data.outputName was enough for that purpose.
> > Otherwise you're gonna laugh but ... I was thinking about it on my lunch time: since we finally don't care about the html-webpack-plugin api and end with a "savage" string replacement, I think there is finally an easier approach which would consist of not relying at all on html-webpack-plugin and just tap the standard `assetEmitted` webpack hook. This way it would be even compatible with any plugins that would produce html files.
> 
> Do you know of any other plugins that do the same thing? I have the impression that this is the most popular. It's true that ideally I'd prefer not to create a link with my plugin because that would require maintenance/doc/monitoring etc.
> 
> If you only use the Webpack API, I don't see how you can detect potential templates associated with entrypoints. We could just replace the hook part extracted from HtmlWebpackPlugin with the native webpack hooks.

Hello @yoriiis 
I meant something in a way like this 
https://github.com/drolsen/webpack-svg-spritely/blob/master/index.js#L285
